File: SRM 651/RobotOnMoon.java
import java.io.*;
import java.util.*;

public class RobotOnMoon {
    private static final int[] dx = new int[]{1, 0, -1, 0};
    private static final int[] dy = new int[]{0, -1, 0, 1};
    private int r, c;
    private int sx, sy;
    private String[] g;
    private boolean in(int x, int y) {
        return x >= 0 && x < r && y >= 0 && y < c;
    }
    private int check(int dir) {
        int curX = sx;
        int curY = sy;
        int len = 0;
        while (true) {
            int nX = curX + dx[dir];
            int nY = curY + dy[dir];
            if (!in(nX, nY)) {
                break;
            }
            if (g[nX].charAt(nY) == '#') {
                return -1;
            }
            ++len;
            curX = nX;
            curY = nY;
        }
        return len;
    }
	public int longestSafeCommand(String[] board) {
        r = board.length;
        c = board[0].length();
        g = board;
        for (int i = 0; i < r; ++i) for (int j = 0; j < c; ++j) {
            if (g[i].charAt(j) == 'S') {
                sx = i;
                sy = j;
            }
        }
        int res[] = new int[2];
        for (int dir = 0; dir < 2; dir++) {
            int len1 = check(dir);
            int len2 = check(dir + 2);
            if (len1 == -1 || len2 == -1) {
                return -1;
            }
            res[dir] = Math.max(res[dir], len1 + len2);
        }
		return res[0] + res[1];
	}

// CUT begin
	public static void main(String[] args){
		System.err.println("RobotOnMoon (250 Points)");
		System.err.println();
		HashSet<Integer> cases = new HashSet<Integer>();
        for (int i = 0; i < args.length; ++i) cases.add(Integer.parseInt(args[i]));
        runTest(cases);
	}

	static void runTest(HashSet<Integer> caseSet) {
	    int cases = 0, passed = 0;
	    while (true) {
	    	String label = Reader.nextLine();
	    	if (label == null || !label.startsWith("--"))
	    		break;

            String[] board = new String[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < board.length; ++i)
                board[i] = Reader.nextLine();
            Reader.nextLine();
            int __answer = Integer.parseInt(Reader.nextLine());

            cases++;
            if (caseSet.size() > 0 && !caseSet.contains(cases - 1))
                continue;
    		System.err.print(String.format("  Testcase #%d ... ", cases - 1));

            if (doTest(board, __answer))
                passed++;
	    }
	    if (caseSet.size() > 0) cases = caseSet.size();
        System.err.println(String.format("%nPassed : %d/%d cases", passed, cases));

        int T = (int)(System.currentTimeMillis() / 1000) - 1500804128;
        double PT = T / 60.0, TT = 75.0;
        System.err.println(String.format("Time   : %d minutes %d secs%nScore  : %.2f points", T / 60, T % 60, 250 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}

	static boolean doTest(String[] board, int __expected) {
		for (int i = 0; i < board.length; i++) {
			board[i] = new String(board[i]);
		}
		long startTime = System.currentTimeMillis();
		Throwable exception = null;
		RobotOnMoon instance = new RobotOnMoon();
		int __result = 0;
		try {
			__result = instance.longestSafeCommand(board);
		}
		catch (Throwable e) { exception = e; }
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
			return false;
		}
		else if (__result == __expected) {
			System.err.println("PASSED! " + String.format("(%.2f seconds)", elapsed));
			return true;
		}
		else {
			System.err.println("FAILED! " + String.format("(%.2f seconds)", elapsed));
			System.err.println("           Expected: " + __expected);
			System.err.println("           Received: " + __result);
			return false;
		}
	}

	static class Reader {
        private static final String dataFileName = "RobotOnMoon.sample";
	    private static BufferedReader reader;

	    public static String nextLine() {
	        try {
                if (reader == null) {
                    reader = new BufferedReader(new InputStreamReader(new FileInputStream(dataFileName)));
                }
                return reader.readLine();
	        }
	        catch (IOException e) {
	            System.err.println("FATAL!! IOException");
	            e.printStackTrace();
	            System.exit(1);
	        }
	        return "";
	    }
	}
// CUT end
}

File: SRM 651/RobotOnMoon.cpp
#include <cstdio>
#include <cmath>
#include <cstring>
#include <ctime>
#include <iostream>
#include <algorithm>
#include <set>
#include <vector>
#include <sstream>
#include <typeinfo>
#include <fstream>

using namespace std;
int dx[] = {1, 0, -1, 0};
int dy[] = {0, 1, 0, -1};
int r, c;
int sx, sy;
vector<string> g;
int in(int x, int y) {
    return x >= 0 && x < r && y >= 0 && y < c;
}
bool check(int dir, int ch, int&len) {
    int cur_x = sx, cur_y = sy;
    len = 0;
    while (true) {
        int nx = cur_x + dx[dir];
        int ny = cur_y + dy[dir];
        if (!in(nx,ny)) break;
        if (g[nx][ny] == ch) return true;
        ++len;
        cur_x = nx;
        cur_y = ny;
    }
    return false;
}
class RobotOnMoon {
    public:
    int longestSafeCommand(vector<string> board) {
        r = board.size();
        c = board[0].size();
        g = board;
        for (int i = 0; i < r; ++i) for (int j = 0; j < c; ++j) {
            if (g[i][j] == 'S') {
                sx = i, sy = j;
            }
        }
        int res[2] = {0};
        for (int i = 0; i < 4; i++) {
            int len1 = 0;
            if (check(i, '#', len1)) {
                return -1;
            }
            int len2 = 0;
            check((i+2)%4, '#', len2);
            res[i%2] = max(res[i%2], len1 + len2);
        }

        return res[0] + res[1];
    }
};

// CUT begin
ifstream data("RobotOnMoon.sample");

string next_line() {
    string s;
    getline(data, s);
    return s;
}

template <typename T> void from_stream(T &t) {
    stringstream ss(next_line());
    ss >> t;
}

void from_stream(string &s) {
    s = next_line();
}

template <typename T> void from_stream(vector<T> &ts) {
    int len;
    from_stream(len);
    ts.clear();
    for (int i = 0; i < len; ++i) {
        T t;
        from_stream(t);
        ts.push_back(t);
    }
}

template <typename T>
string to_string(T t) {
    stringstream s;
    s << t;
    return s.str();
}

string to_string(string t) {
    return "\"" + t + "\"";
}

bool do_test(vector<string> board, int __expected) {
    time_t startClock = clock();
    RobotOnMoon *instance = new RobotOnMoon();
    int __result = instance->longestSafeCommand(board);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << to_string(__expected) << endl;
        cout << "           Received: " << to_string(__result) << endl;
        return false;
    }
}

int run_test(bool mainProcess, const set<int> &case_set, const string command) {
    int cases = 0, passed = 0;
    while (true) {
        if (next_line().find("--") != 0)
            break;
        vector<string> board;
        from_stream(board);
        next_line();
        int __answer;
        from_stream(__answer);

        cases++;
        if (case_set.size() > 0 && case_set.find(cases - 1) == case_set.end())
            continue;

        cout << "  Testcase #" << cases - 1 << " ... ";
        if ( do_test(board, __answer)) {
            passed++;
        }
    }
    if (mainProcess) {
        cout << endl << "Passed : " << passed << "/" << cases << " cases" << endl;
        int T = time(NULL) - 1500855799;
        double PT = T / 60.0, TT = 75.0;
        cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
        cout << "Score  : " << 250 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
    set<int> cases;
    bool mainProcess = true;
    for (int i = 1; i < argc; ++i) {
        if ( string(argv[i]) == "-") {
            mainProcess = false;
        } else {
            cases.insert(atoi(argv[i]));
        }
    }
    if (mainProcess) {
        cout << "RobotOnMoon (250 Points)" << endl << endl;
    }
    return run_test(mainProcess, cases, argv[0]);
}
// CUT end
